### PR TITLE
Fix/12-implement-kg (#12)

### PR DESCRIPTION
**Pull request by msesboue**
Fix: change graph loader with knowledge graph class and update it.

Run only the knowledge graph tests. I did not want to remove the entity matcher tests yet but they are not relevant anymore.

### CODE REVIEW

File: buzz_el/graph_loader/knowledge_graph.py [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+        Tag specifying how to interpret the KG file, by default "RDF".
+
+    Attributes
+    ----------
+    _kg_file_path : PathLike
+        The path to the knowledge graph file.
+    annotation_properties : Optional[Set[str]], optional
+        Set of RDF annotation properties used to extract entity labels, by default {"rdfs:label"}.
+    _annotation_properties: Set[str]
+        Annotation properties processed to be used directly in a SPARQL query.
```
Comment (schmarion) [inline buzz_el/graph_loader/knowledge_graph.py:32]: why 2 times?

File: buzz_el/graph_loader/knowledge_graph.py [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+    _lang_filter : Optional[str], optional
+        Language filter tag to filter entity labels based on language, by default None.
+    _sparql_alternative_path_str: str
+        The portion of the SPARQL query constituting the alternative paths. 
+    _sparql_lang_filter_str: str
+        The portion of the SPARQL query constituting the language filter.
```
Comment (schmarion) [inline buzz_el/graph_loader/knowledge_graph.py:38]: why 2 times ?

File: buzz_el/graph_loader/knowledge_graph.py [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+            Set of RDF annotation properties used to extract entity labels, by default {"rdfs:label"}.
+        lang_filter_tag : Optional[str], optional
+            Language filter tag to filter entity labels based on language, by default None.
+        interpretation_type : Optional[str], optional
+           Tag specifying how to interpret the KG file, by default "RDF".
+        """
+        self._kg_file_path = ensure_path(kg_file_path)
+
+        self._annotation_properties = {"rdfs:label"}
+        self.annotation_properties = annotation_properties
```
Comment (schmarion) [inline buzz_el/graph_loader/knowledge_graph.py:69]: I suppose that you are doing this to use the setter but I don't understand when the self._annotation_properties is set with self.annotation_properties

File: buzz_el/graph_loader/knowledge_graph.py [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+
+        self._lang_filter = lang_filter_tag
+        self._sparql_lang_filter_str = (
+            f'FILTER ( lang(?label) = "{self._lang_filter}" )'
+            if self._lang_filter
+            else ""
+        )
+
+        self._interpretation_type = None
+        self.interpretation_type = interpretation_type
```
Comment (schmarion) [inline buzz_el/graph_loader/knowledge_graph.py:81]: I suppose that you are doing this to use the setter but I don't understand when the `self._interpretation_type` is set with self.interpretation_type

File: buzz_el/graph_loader/knowledge_graph.py [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+                else:
+                    annotation_properties.add(prop)
+
+            self._annotation_properties = annotation_properties
+
+    def load_kg_from_file(self) -> None:
+        """Load the knowledge graph from the specified file."""
+        self.kg.parse(self._kg_file_path)
+
+    def sparql_endpoint(self, sparql_query: str) -> Iterable:
```
Comment (schmarion) [inline buzz_el/graph_loader/knowledge_graph.py:152]: Why this is an Iterable and not a string ?

File: buzz_el/graph_loader/knowledge_graph.py [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+        for res in sparql_res:
+            patterns.append({
+                "label": "KG_ENT", 
+                "pattern": str(res["label"]),
+                "id": str(res["ent_uri"])
+            })
+
+        return patterns
+
+    def get_context(self, entity_uri: str) -> List[str]:
```
Comment (schmarion) [inline buzz_el/graph_loader/knowledge_graph.py:214]: How can we customize that in the future ? 
I am not a big fan of the implementation. 
We will have different  `_build_ent_context_sparql_query` function ? We have to add a callable as parameter ?

File: examples/data/pizzas_bisou_sample.ttl [hunk 7]
```diff
@@ -0,0 +1,103 @@
+@prefix pizza: <http://www.msesboue.org/o/pizza-data-demo/bisou#> .
```
[the remark this answers is not in the document]
Comment (msesboue) [reply examples/data/pizzas_bisou_sample.ttl:1]: I would say yes. And nothing really link them there.

File: examples/entity_matcher_pizza_kg_demo.ipynb [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import spacy\n",
+    "from random import sample\n",
+    "\n",
+    "from buzz_el.graph_loader import GraphLoader\n",
```
[the remark this answers is not in the document]
Comment (msesboue) [reply examples/entity_matcher_pizza_kg_demo.ipynb:25]: I will do this in another MR

File: buzz_el/graph_loader/knowledge_graph.py (hunk 1 above)
Comment (msesboue) [reply buzz_el/graph_loader/knowledge_graph.py]: The private version adds up some modifications so I can use the strings directly in the SPARQL queries. If there is a namespace I leave it as it is. But if it is a URI, I need to add the "<URI>".

File: buzz_el/graph_loader/knowledge_graph.py (hunk 3 above)
Comment (msesboue) [reply buzz_el/graph_loader/knowledge_graph.py]: When you do `self.annotation_properties = annotation_properties`, you are actually calling the setter, which sets `self._annotation_properties`.

Remember, a while back I played around with this (maybe too much, but I believe it is a good occasion)

File: buzz_el/graph_loader/knowledge_graph.py (hunk 4 above)
Comment (msesboue) [reply buzz_el/graph_loader/knowledge_graph.py]: same as above

File: buzz_el/graph_loader/knowledge_graph.py (hunk 6 above)
Comment (msesboue) [reply buzz_el/graph_loader/knowledge_graph.py]: That is a great question. I had the same, but I wanted to implement a first version. It should definitely change in the future.

**Comment by msesboue** [comment]
I made the discussed changes, i.e., created a KG and graph loader objects.

I believe the `.vscode` file has been updated due to a VScode update.

Potential unexpected dependency:

How we designed the graph loader and KG classes might have introduced a dependency between them. When constructing the get context method, we use some SPARQL query creation methods defined in the graph loader. Hence, I think if we deleted the graph loader after generating the KG, the KG get context method might be broken. In other words, the graph loader does not create a standalone KG instance. It might be an issue in the future. I don't know.

File: buzz_el/graph/graph_loader.py [hunk 9]
```diff
@@ -0,0 +1,307 @@
...
+            SELECT DISTINCT ?{self._sparql_var} WHERE {{
+                <{entity_uri}> {self._context_sparql_alt_path_str} ?{self._sparql_var} .
+                {self._sparql_lang_filter_str}
+            }}
+        """
+
+        return ent_context_query
+
+    def kg_get_context(self) -> Callable[[str], str]:
+        """Build and return the knowledge graph instance get context method.
```
Comment (schmarion) [inline buzz_el/graph/graph_loader.py:261]: get_context method

File: buzz_el/graph/graph_loader.py [hunk 10]
```diff
@@ -0,0 +1,307 @@
...
+
+        Returns
+        -------
+        Callable[[str], str]
+            The get context method.
```
Comment (schmarion) [inline buzz_el/graph/graph_loader.py:266]: _The get_context method._ or _The method to get an entity context._

File: buzz_el/graph/graph_loader.py [hunk 11]
```diff
@@ -0,0 +1,307 @@
...
+
+        self._context_properties = None
+        self.context_properties = context_properties
+
+        self._label_sparql_alt_path_str = "|".join(self._label_properties)
+
+        if self._context_properties is None:
+            self._context_sparql_alt_path_str = None
+        else:
+            self._context_sparql_alt_path_str = "|".join(self._context_properties)
```
[the remark this answers is not in the document]
Comment (msesboue) [reply buzz_el/graph/graph_loader.py:93]: same response

File: buzz_el/graph/graph_loader.py [hunk 12]
```diff
@@ -0,0 +1,307 @@
...
+
+        self._sparql_var = "label"
```
[the remark this answers is not in the document]
Comment (msesboue) [reply buzz_el/graph/graph_loader.py:95]: The get context method relies on the name of this variable (line 279), which is defined in both queries. We can give it the name we like, but it has to be the same in all SPARQL queries.

**Pull request merged**